query-engine/driver-adapters/src/proxy.rs

```diff
@@ -581,34 +580,50 @@ impl TransactionProxy {
     pub fn new(js_transaction: &JsObject) -> napi::Result<Self> {
         let commit = js_transaction.get_named_property("commit")?;
         let rollback = js_transaction.get_named_property("rollback")?;
-        let dispose = js_transaction.get_named_property("dispose")?;
         let options = js_transaction.get_named_property("options")?;
 
         Ok(Self {
             commit,
             rollback,
-            dispose,
             options,
+            closed: AtomicBool::new(false),
         })
     }
 
     pub fn options(&self) -> &TransactionOptions {
         &self.options
     }
 
+    /// Commits the transaction via the driver adapter.
+    ///
+    /// Cancelation safety: [`TransactionProxy::closed`] is only set after the call to JS finishes,
+    /// so the destructor will ensure the transaction is closed even if the future is dropped.
     pub async fn commit(&self) -> quaint::Result<()> {
-        self.commit.call(()).await
+        let result = self.commit.call(()).await;
+        self.closed.swap(true, Ordering::Release);
+        result
     }
 
+    /// Rolls back the transaction via the driver adapter.
+    ///
+    /// Cancelation safety: [`TransactionProxy::closed`] is only set after the call to JS finishes,
+    /// so the destructor will ensure the transaction is closed even if the future is dropped.
     pub async fn rollback(&self) -> quaint::Result<()> {
-        self.rollback.call(()).await
+        let result = self.rollback.call(()).await;
+        self.closed.swap(true, Ordering::Release);
+        result
     }
 }
 
 impl Drop for TransactionProxy {
     fn drop(&mut self) {
+        if self.closed.swap(true, Ordering::Acquire) {
+            return;
+        }
+
         _ = self
-            .dispose
+            .rollback
+            .as_raw()
             .call((), napi::threadsafe_function::ThreadsafeFunctionCallMode::NonBlocking);
```

Comment:
The issue suggests spawning a new tokio task and awaiting the result which is also possible of course:
```diff
diff --git a/query-engine/driver-adapters/src/async_js_function.rs b/query-engine/driver-adapters/src/async_js_function.rs
index 4926534f58..c6cb52def7 100644
--- a/query-engine/driver-adapters/src/async_js_function.rs
+++ b/query-engine/driver-adapters/src/async_js_function.rs
@@ -17,6 +17,7 @@ use crate::{
 /// - Unpacks JS `Result` type into Rust `Result` type and converts the error
 /// into `quaint::Error`.
 /// - Catches panics and converts them to `quaint:Error`
+#[derive(Clone)]
 pub(crate) struct AsyncJsFunction<ArgType, ReturnType>
 where
     ArgType: ToNapiValue + 'static,
@@ -55,10 +56,6 @@ where
         .map_err(into_quaint_error)?;
         js_result.into()
     }
-
-    pub(crate) fn as_raw(&self) -> &ThreadsafeFunction<ArgType, ErrorStrategy::Fatal> {
-        &self.threadsafe_fn
-    }
 }
 
 impl<ArgType, ReturnType> FromNapiValue for AsyncJsFunction<ArgType, ReturnType>
diff --git a/query-engine/driver-adapters/src/proxy.rs b/query-engine/driver-adapters/src/proxy.rs
index 000203c0d2..b1512dc618 100644
--- a/query-engine/driver-adapters/src/proxy.rs
+++ b/query-engine/driver-adapters/src/proxy.rs
@@ -621,10 +621,11 @@ impl Drop for TransactionProxy {
             return;
         }
 
-        _ = self
-            .rollback
-            .as_raw()
-            .call((), napi::threadsafe_function::ThreadsafeFunctionCallMode::NonBlocking);
+        let rollback = self.rollback.clone();
+
+        napi::tokio::spawn(async move {
+            _ = rollback.call(()).await;
+        });
     }
 }
 ```

but the current implementation in this PR is also fine as long as the exceptions in `rollback` are guaranteed to be caught on JS side, and feels better to me.

Reply:
let me know if you'd prefer that instead though

Reply:
I think it's ok the way it is.